ArdusatSDK.cpp

```diff
@@ -453,266 +257,1212 @@ const char * valuesToCSV(const char *sensorName, unsigned long timestamp, int nu
+    return valuesToCSV(sensorName, Magnetic::header.timestamp, 3,
+                       Magnetic::x, Magnetic::y, Magnetic::z);
+  else
+    return "";
+}
+
+/**
+ * @brief   Returns last read value in JSON format
+ * @ingroup magnetic
+ * @param   sensorName The text to display next to the value
+ */
+const char * Magnetic::toJSON(const char * sensorName) {
+  if (Magnetic::initialized)
+    return _tripleValueToJSON(sensorName, 2, Magnetic::header.unit, "%sX", Magnetic::x,
+                             "%sY", Magnetic::y, "%sZ", Magnetic::z);
+  else
+    return "";
+}
+
+
+/**************************************************************************//**
+ * @brief   Constructs Orientation calculation object
+ * @ingroup orientation
+ *
+ * Example Usage:
+ * @code
+ *     Orientation orient = Orientation();  // Instantiate sensor object
+ *     orient.begin();                      // Initialize sensor
+ *     Serial.println(orient.readToJSON()); // Read and print values in JSON
+ * @endcode
+ *****************************************************************************/
+Orientation::Orientation(void) {
+  Orientation::sensorId = SENSORID_ADAFRUIT9DOFIMU;
+  Orientation::header.unit = DATA_UNIT_DEGREES;
+  Orientation::initialized = false;
+}
+
+Orientation::~Orientation() {
+}
+
+/**
+ * @brief   Creates new Acceleration and Magnetic Sensors to derive orientation
+ * @ingroup orientation
+ *
+ * @retval true  Successfully initialized
+ * @retval false Failed to initialize
+ */
+boolean Orientation::begin(void) {
+  Acceleration accel;
+  Magnetic mag;
+  accel.begin();
+  mag.begin();
+  return Orientation::begin(accel, mag);
+}
+
+/**
+ * @brief   Uses provided Acceleration and Magnetic Sensors to derive orientation
+ * @ingroup orientation
+ *
+ * @param accel Existing Acceleration Sensor to use
+ * @param mag   Existing Magnetic Sensor to use
+ *
+ * @retval true  Successfully initialized
+ * @retval false Failed to initialize
+ */
+boolean Orientation::begin(Acceleration & accel, Magnetic & mag) {
+  catchSpaceboard();
+  Orientation::accel = & accel;
+  Orientation::mag = & mag;
+  Orientation::initialized = accel.initialized && mag.initialized;
+  return Orientation::initialized;
+}
+
+/**
+ * @brief   Takes a reading from the sensor
+ * @ingroup orientation
+ */
+void Orientation::read(void) {
+  Orientation::accel->read();
+  Orientation::mag->read();
+  Orientation::read(*(Orientation::accel), *(Orientation::mag));
+}
+
+/**
+ * @brief   Takes a reading from the sensor
+ * @ingroup orientation
+ *
+ * @param accel Existing Acceleration Sensor to use
+ * @param mag   Existing Magnetic Sensor to use
+ */
+void Orientation::read(Acceleration & acc, Magnetic & mag) {
+  float roll;
+  float pitch;
+  float heading;
+  const float PI_F = 3.141592653F;
+
+  // Roll is rotation around x-axis (-180 <= roll <= 180)
+  // Positive roll is clockwise rotation wrt positive x axis
+  roll = (float) atan2(acc.y, acc.z);
```

Comment:
I intentionally made the `read(accel, mag);` function not call `read()` on the accel or mag objects. I figured it was possible someone might want to calculate the orientation based on previously read accel and mag values. However, there is a `read()` function that does call `read()` on accel and mag individually.